### PR TITLE
Angus/database service

### LINKED ISSUE #911: Decouple AuthN/Z
Want to decouple the Authentication (AuthN) and Authorisation (AuthZ) from CARTA. This will make it easier to plug CARTA into different sites and in particular use CARTA in conduction with pluggable IdP proxies.

### LINKED ISSUE #912: Front end authN/Z decouple
Remove login GUI and modify to interact with pre-authenticated CARTA startup server.

### PR DESCRIPTION
**Pull request by veggiesaurus**
Closes #912, closes #911 

This PR introduces an API service, which is used for preferences or layouts in an agnostic manner. For the server version, HTTP requests are used. For the desktop application (or simple server installations without authentication), local storage is used, preferences are stored as a single JSON string and validated against a JSON schema using AJV, removing the need for manual validator functions. Layouts are stored in the same way as before, but are validated against a much more comprehensive JSON schema, again removing the need for manual validation. 

When encountering a local storage without the `preferences` item, this item is automatically created, and older versions of the preferences (in separate untyped key-value pairs) are upgraded to this new approach.

This PR also adopts the new authentication approach, which allows us to remove all mongo and JWT behaviour from the backend itself.

### CODE REVIEW

**Comment by veggiesaurus** [comment]
@kswang1029 careful testing of this branch is required when it comes to upgrading existing preferences correctly.

**Comment by kswang1029** [comment]
@veggiesaurus not sure I tested it right but I will try to describe it.

1) using the "server" with google login that @ajm-asiaa set up before, I tried to make a custom layout and set it as my default layout. In addition, I tried to set several custom preferences in the preferences dialogue. Then @ajm-asiaa helped to upgrade the frontend to aec75aa. Once I re-login again and launch the updated frontend, I see
Screen Shot 2020-08-04 at 9 48 56 AM
and all those custom preferences are not recovered. Do we expect we see all the preferences and the layout recovered ("upgraded") in this case?

2) using the mix page, I did similar with the frontend dev branch then switch to aec75aa, I did not see my preferences and the custom default layout recovered. Do we expect we see all the preferences and the layout recovered ("upgraded") in this case, too?

**Comment by kswang1029** [comment]
@veggiesaurus when using the desktop version (or simple server without auth) do we need internet access to validate the JSON string (e.g. traveling without wifi)?

**Comment by ajm-ska** [comment]
@veggiesaurus I updated this `angus/database_server` frontend branch on the server where we are running carta-node-server so that @kswang1029 could test it. At first I had trouble starting up carta-node-server again. The error message was saying 

> 
> Serving CARTA frontend from /etc/carta/frontend
> MongoError: a collection 'CARTA.preferences' already exists
>     at MessageStream.messageHandler (/home/acdc/.npm/_npx/302890/lib/node_modules/carta-node-server/node_modules/mongodb/lib/cmap/connection.js:266:20)
>     at MessageStream.emit (events.js:315:20)
>     at processIncomingData (/home/acdc/.npm/_npx/302890/lib/node_modules/carta-node-server/node_modules/mongodb/lib/cmap/message_stream.js:144:12)
>     at MessageStream._write (/home/acdc/.npm/_npx/302890/lib/node_modules/carta-node-server/node_modules/mongodb/lib/cmap/message_stream.js:42:5)
>     at doWrite (_stream_writable.js:403:12)
>     at writeOrBuffer (_stream_writable.js:387:5)
>     at MessageStream.Writable.write (_stream_writable.js:318:11)
>     at Socket.ondata (_stream_readable.js:717:22)
>     at Socket.emit (events.js:315:20)
>     at addChunk (_stream_readable.js:295:12) {
>   ok: 0,
>   code: 48,
>   codeName: 'NamespaceExists'
> }
> Error connecting to database
> 

I guess there is a conflict with the earlier versions of whatever is saved to the MongoDB and that wouldn't be a problem in the final release version. So I entered `mongo` and deleted the `CARTA` database (mongo, show dbs, use CARTA, db.dropDatabase(), exit() )

Then carta-node-server starts up fine.

But when I logged in to the frontend, I saw this message:
Screenshot 2020-08-04 at 09 42 22
I can click OK to make it go away. It comes up if I log out and in again.

I don't seem to be able to save user defined layouts either:
Screenshot 2020-08-04 at 09 58 24

Perhaps it is a permissions issue somewhere? It is strange as `CARTA` entry appears to have been regenerated in the MongoDB.

**Comment by veggiesaurus** [comment]
> @veggiesaurus I updated this `angus/database_server` frontend branch on the server where we are running carta-node-server so that @kswang1029 could test it. At first I had trouble starting up carta-node-server again. The error message was saying 
> 
> > 
> > Serving CARTA frontend from /etc/carta/frontend
> > MongoError: a collection 'CARTA.preferences' already exists
> >     at MessageStream.messageHandler (/home/acdc/.npm/_npx/302890/lib/node_modules/carta-node-server/node_modules/mongodb/lib/cmap/connection.js:266:20)
> >     at MessageStream.emit (events.js:315:20)
> >     at processIncomingData (/home/acdc/.npm/_npx/302890/lib/node_modules/carta-node-server/node_modules/mongodb/lib/cmap/message_stream.js:144:12)
> >     at MessageStream._write (/home/acdc/.npm/_npx/302890/lib/node_modules/carta-node-server/node_modules/mongodb/lib/cmap/message_stream.js:42:5)
> >     at doWrite (_stream_writable.js:403:12)
> >     at writeOrBuffer (_stream_writable.js:387:5)
> >     at MessageStream.Writable.write (_stream_writable.js:318:11)
> >     at Socket.ondata (_stream_readable.js:717:22)
> >     at Socket.emit (events.js:315:20)
> >     at addChunk (_stream_readable.js:295:12) {
> >   ok: 0,
> >   code: 48,
> >   codeName: 'NamespaceExists'
> > }
> > Error connecting to database
> > 
> 
> I guess there is a conflict with the earlier versions of whatever is saved to the MongoDB and that wouldn't be a problem in the final release version. So I entered `mongo` and deleted the `CARTA` database (mongo, show dbs, use CARTA, db.dropDatabase(), exit() )
> 
> Then carta-node-server starts up fine.
> 
> But when I logged in to the frontend, I saw this message:
> Screenshot 2020-08-04 at 09 42 22
> I can click OK to make it go away. It comes up if I log out and in again.
> 
> I don't seem to be able to save user defined layouts either:
> Screenshot 2020-08-04 at 09 58 24
> 
> Perhaps it is a permissions issue somewhere? It is strange as `CARTA` entry appears to have been regenerated in the MongoDB.
> 

I haven't implemented the server side yet. I'll do that today.

**Comment by veggiesaurus** [comment]
> @veggiesaurus when using the desktop version (or simple server without auth) do we need internet access to validate the JSON string (e.g. traveling without wifi)?

No, it's all done by the frontend

**Comment by veggiesaurus** [comment]
> 2) using the mix page, I did similar with the frontend dev branch then switch to aec75aa, I did not see my preferences and the custom default layout recovered. Do we expect we see all the preferences and the layout recovered ("upgraded") in this case, too?

The upgrade is once off, so it should only have upgraded your initial settings you had stored before accessing this branch. Did it do that? Can you look in the JavaScript console to see what is in your local storage? There should be a preferences item

I'll update the server branch today... Perhaps best to avoid testing that for now. Sorry, I should have mentioned that first 🙈

**Comment by kswang1029** [comment]
I see my custom preferences (eg layout=ksw_default2) but all these custom preferences are not applied to the GUI.

Screen Shot 2020-08-04 at 1 34 10 PM

**Comment by veggiesaurus** [comment]
> I see my custom preferences (eg layout=ksw_default2) but all these custom preferences are not applied to the GUI.
> 
> Screen Shot 2020-08-04 at 1 34 10 PM
> 

Those are the old preferences. The upgraded preferences should be under a single item called preferences. Is that there?

**Comment by kswang1029** [comment]
@veggiesaurus I only see one item under local storage (not "preferences").

Screen Shot 2020-08-04 at 1 40 35 PM

**Comment by veggiesaurus** [comment]
> @veggiesaurus I only see one item under local storage (not "preferences").
> 
> Screen Shot 2020-08-04 at 1 40 35 PM
> 

Sorry, I mean inside that entry, is there a key called "preferences"?

**Comment by kswang1029** [comment]
oh... yes, there is! but the contents are not "upgraded" to my custom one (e.g. layout="Default", should be "ksw_default2")
Screen Shot 2020-08-04 at 1 43 43 PM

**Comment by kswang1029** [comment]
OK, after some guidelines from @veggiesaurus, it works! (was accidentally accessing this branch _before_ my test so actually the preferences did being upgraded)

**Comment by duidae** [comment]
looked some of the codes related to layout, it seems that we are only checking the config of version 2, so @veggiesaurus are we going to obsolete version 1 completely?

**Comment by veggiesaurus** [comment]
> looked some of the codes related to layout, it seems that we are only checking the config of version 2, so @veggiesaurus are we going to obsolete version 1 completely?

This schema should support version 1 as well. Isn't it backwards compatible? The only difference I could see is the types of widgets available, and the X coordinate added to the spatial profiler settings (which I've taken care of using the default value in the schema)

**Comment by duidae** [comment]
I dont see the part of schema handling the property of "coord" (not "coordinate") of version 1, that is the minor difference between ver 1 & ver 2.

**Comment by duidae** [comment]
btw I'll suggest we can start to group those the settings in preference schema for extensibility(maybe not now), such as "theme", "autolaunch" under "global", so that we wont have to give long property names and is able to avoid key conflict in the future.

**Comment by veggiesaurus** [comment]
> I dont see the part of schema handling the property of "coord" (not "coordinate") of version 1, that is the minor difference between ver 1 & ver 2.

have updated to include v1->v2 upgrading

**Comment by veggiesaurus** [comment]
> btw I'll suggest we can start to group those the settings in preference schema for extensibility(maybe not now), such as "theme", "autolaunch" under "global", so that we wont have to give long property names and is able to avoid key conflict in the future.

I think this is a good idea, but the only issue I have with this is that it couples the config to the UI itself. If we decide to move a preference from one tab to another, it would require a schema change.

It would be a significant change, though, and perhaps its best to leave this for a preferences v2 schema

**Comment by duidae** [comment]
ver1 would be like:

"type":"component",
"id":"spatial-profiler",
"coord":"x"


ver2 would be like:

"type":"component",
"id":"spatial-profiler",
"widgetSettings":{
   "coordinate":"x",
   ...
}

**Review by duidae — CHANGES_REQUESTED** [review]
ver1 would be like:

"type":"component",
"id":"spatial-profiler",
"coord":"x"

============================

ver2 would be like:

"type":"component",
"id":"spatial-profiler",
"widgetSettings":{
"coordinate":"x",
...
}

File: src/models/LayoutConfig.ts [hunk 1]
```diff
@@ -112,6 +112,19 @@ export class LayoutConfig {
         };
     };
 
+    public static UpgradeLayout = (layout: { layoutVersion: 1 | 2, docked: any, floating: any }) => {
+        // Upgrade to V2 if required
+        if (layout.layoutVersion === 1) {
+            const spatialProfileWidgets = findDeep(layout, item => item.id === "spatial-profiler");
+            for (const widget of spatialProfileWidgets) {
+                if (widget.widgetSettings?.coord) {
```
Comment (duidae) [inline src/models/LayoutConfig.ts]: coord is not nested inside widgetSettings in ver1

Comment (veggiesaurus) [reply src/models/LayoutConfig.ts]: oops, ok. Will fix now :)

**Review by duidae — APPROVED** [review]
looks good, btw yesterday I downloaded the appImage of ver 1.2 & tried to check the saved config in localstorage but find no way, wondering is there a way to do that or should we activate the dev tools of electron for packaged CARTA binary? Because once there are bugs related to those things, we may have no way to debug since users cant open the console and provide anything.

**Comment by kswang1029** [comment]
@duidae there is a hacky way (modifying main.js?) to enable dev tools. @ajm-asiaa should know how to.

**Comment by veggiesaurus** [comment]
> looks good, btw yesterday I downloaded the appImage of ver 1.2 & tried to check the saved config in localstorage but find no way, wondering is there a way to do that or should we activate the dev tools of electron for packaged CARTA binary? Because once there are bugs related to those things, we may have no way to debug since users cant open the console and provide anything.

I think @ajm-asiaa added a `--debug` flag to the 1.4 alpha. Is that right?

Don't they all share the same local storage, though? Or did we only unify them at 1.3?

**Comment by ajm-ska** [comment]
@duidae @veggiesaurus Correct. The v1.4 Electron Desktop versions will have a flag to open to DevTools:

> [--debug] Open the DevTools in the Electron window.

To do so in the current and older AppImage versions, you need to: 
1. Extract the AppImage: `./CARTA.AppImage --appimage-extract`
2. Open the main.js file: `squashfs-root/resources/app/main.js`
3. Manually comment in the line with: `mainWindow.webContents.openDevTools()`.
4. Start CARTA up from the extracted folder `squashfs-root/AppRun`

**Comment by duidae** [comment]
> @duidae @veggiesaurus Correct. The v1.4 Electron Desktop versions will have a flag to open to DevTools:
> 
> > [--debug] Open the DevTools in the Electron window.
> 
> To do so in the current and older AppImage versions, you need to:
> 
> 1. Extract the AppImage: `./CARTA.AppImage --appimage-extract`
> 2. Open the main.js file: `squashfs-root/resources/app/main.js`
> 3. Manually comment in the line with: `mainWindow.webContents.openDevTools()`.
> 4. Start CARTA up from the extracted folder `squashfs-root/AppRun`

that's great, @ajm-asiaa  thanks!

**Pull request merged**